### PR TITLE
Create todo-mvp-tablet with a tablet presenter

### PR DESCRIPTION
**Pull request by JoseAlcerreca**
Also creates a TasksMvpController, TasksTabletNavigator, removes BasePresenter and bumps versions

### CODE REVIEW

File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/TasksMvpController.java [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+        mFragmentActivity = fragmentActivity;
+    }
+
+    private void initTasksView() {
+        if (isTablet()) {
+            createTabletElements();
+        } else {
+            TasksFragment tasksFragment = findOrCreateTasksFragment();
+            if (tasksFragment.getPresenter() != null) {
+                mTasksPresenter = (TasksPresenter) tasksFragment.getPresenter();
```
Comment (malmstein) [inline todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/TasksMvpController.java]: how about letting the `Controller` deal with this? the `Controller` could be attached to the `Activity` lifecycle and would be in charge of assigning the `Presenter` to each `Fragment`

File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/TasksMvpController.java [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+    public TaskDetailPresenter getTaskDetailPresenter() {
+        return mTaskDetailPresenter;
+    }
+
+    /**
+     * To be called in tablet mode when a new detail view is requested.
+     * @param taskId the ID of the task
+     * @return the fragment to be added to the UI
+     */
+    public TaskDetailFragment createDetailViewForTablet(String taskId) {
```
Comment (malmstein) [inline todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/TasksMvpController.java]: this is called from a `Navigator` class, I don't think it's its responsibility to create `Fragments`. The reason why it's called is because instead of refreshing the `Task` that is being shown we simple recreate the `Fragment`. What do you think about making the `DetailFragment` able to refresh the task it's displaying?

File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/TasksMvpController.java [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+        return tasksMvpController;
+    }
+
+    // Force factory method, prevent direct instantiation:
+    private TasksMvpController(@NonNull FragmentActivity fragmentActivity) {
+        mFragmentActivity = fragmentActivity;
+    }
+
+    private void initTasksView() {
+        if (isTablet()) {
```
Comment (malmstein) [inline todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/TasksMvpController.java]: I like this, the `Controller` is in charge to setup `Tablet` or not.

File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/TasksMvpController.java [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+        mTasksTabletPresenter.setTaskDetailPresenter(taskDetailPresenter);
+        taskDetailFragment.setPresenter(mTasksTabletPresenter);
+        return taskDetailFragment;
+    }
+
+    private void createTabletElements() {
+        TasksFragment tasksFragment = findOrCreateTasksFragment();
+
+        // TasksFragment is retained so let's reuse its presenter.
+        if (tasksFragment.getPresenter() != null) {
```
Comment (malmstein) [inline todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/TasksMvpController.java]: same as above, the `Fragment` should not expose the `Presenter`, I'd rather have this `Controller` in charge of that.

File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/TasksMvpController.java [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+            // Create the fragment
+            tasksFragment = TasksFragment.newInstance();
+            ActivityUtils.addFragmentToActivity(
+                    getSupportFragmentManager(), tasksFragment, R.id.contentFrame);
+        }
+        return tasksFragment;
+    }
+
+    @NonNull
+    private TaskDetailFragment findOrCreateTaskDetailFragment(String taskId) {
```
Comment (malmstein) [inline todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/TasksMvpController.java]: this is the main purpose of the `Controller` 👍

File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/TasksMvpController.java [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+        return (TaskDetailFragment) mFragmentActivity.getSupportFragmentManager().findFragmentById(
+                R.id.contentFrame_detail);
+    }
+
+    private FragmentManager getSupportFragmentManager() {
+        return mFragmentActivity.getSupportFragmentManager();
+    }
+
+    private boolean isTablet() {
+        return mFragmentActivity.getResources().getBoolean(R.bool.isTablet);
```
Comment (malmstein) [inline todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/TasksMvpController.java]: 👍

File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/TasksTabletNavigator.java [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+import android.support.v4.app.Fragment;
+import android.support.v4.app.FragmentActivity;
+import android.support.v4.app.FragmentTransaction;
+
+import com.example.android.architecture.blueprints.todoapp.taskdetail.TaskDetailFragment;
+
+/**
+ * Manages screens in tablet mode.
+ */
+public class TasksTabletNavigator {
```
Comment (malmstein) [inline todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/TasksTabletNavigator.java]: I think that everything here belongs to the `Controller`. The main purpose of a `Navigator` should be navigating between screens

File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/TasksTabletPresenter.java [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+import com.example.android.architecture.blueprints.todoapp.taskdetail.TaskDetailContract;
+import com.example.android.architecture.blueprints.todoapp.taskdetail.TaskDetailPresenter;
+import com.example.android.architecture.blueprints.todoapp.tasks.TasksContract;
+import com.example.android.architecture.blueprints.todoapp.tasks.TasksFilterType;
+import com.example.android.architecture.blueprints.todoapp.tasks.TasksPresenter;
+
+/**
+ * Presenter for the tablet screen that can act as a Tasks Presenter and a Task Detail Presenter.
+ */
+public class TasksTabletPresenter implements TasksContract.Presenter, TaskDetailContract.Presenter {
```
Comment (malmstein) [inline todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/TasksTabletPresenter.java]: I like this, can't think of a better way to handle the `TabletPresenter`.

File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/TasksTabletPresenter.java [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+    }
+
+    @Override
+    public void addNewTask() {
+        mTasksPresenter.addNewTask();
+    }
+
+    @Override
+    public void openTaskDetails(@NonNull Task requestedTask) {
+        mTasksNavigator.startTaskDetailForTablet(requestedTask.getId());
```
Comment (malmstein) [inline todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/TasksTabletPresenter.java]: this is the only thing I'm not sure about, doesn't seem the `Navigator` responsibility to do this. Can we have a reference to the `Controller` instead?

File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskFragment.java [hunk 10]
```diff
@@ -56,7 +56,7 @@ public AddEditTaskFragment() {
     @Override
     public void onResume() {
         super.onResume();
-        mPresenter.start();
+        mPresenter.startAddEditTaskPresenter();
```
Comment (malmstein) [inline todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskFragment.java]: `populateTask()` and this do the same thing, can we get rid of one?

File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/statistics/StatisticsContract.java [hunk 11]
```diff
@@ -35,7 +34,7 @@
         boolean isActive();
     }
 
-    interface Presenter extends BasePresenter {
-
+    interface Presenter {
+        void startStatisticsPresenter();
```
Comment (malmstein) [inline todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/statistics/StatisticsContract.java]: better naming here would be nice, this is later calling `loadStatistics()`. How about changing this to `presenter.loadStatistics()`?

File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/statistics/StatisticsFragment.java [hunk 12]
```diff
@@ -59,7 +59,7 @@ public View onCreateView(LayoutInflater inflater, ViewGroup container,
     @Override
     public void onResume() {
         super.onResume();
-        mPresenter.start();
+        mPresenter.startStatisticsPresenter();
```
Comment (malmstein) [inline todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/statistics/StatisticsFragment.java]: see comments above

File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/statistics/StatisticsPresenter.java [hunk 13]
```diff
@@ -46,7 +46,7 @@ public StatisticsPresenter(@NonNull TasksRepository tasksRepository,
     }
 
     @Override
-    public void start() {
+    public void startStatisticsPresenter() {
```
Comment (malmstein) [inline todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/statistics/StatisticsPresenter.java]: as per comments above

File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/taskdetail/TaskDetailContract.java [hunk 14]
```diff
@@ -60,5 +59,7 @@
         void completeTask();
 
         void activateTask();
+
+        void startTaskDetailPresenter();
```
Comment (malmstein) [inline todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/taskdetail/TaskDetailContract.java]: how about calling this `openTask()`?

File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/taskdetail/TaskDetailFragment.java [hunk 15]
```diff
@@ -77,20 +77,26 @@ public View onCreateView(LayoutInflater inflater, ViewGroup container,
                              Bundle savedInstanceState) {
         View root = inflater.inflate(R.layout.taskdetail_frag, container, false);
         setHasOptionsMenu(true);
+
+        // Retain the fragment to persist the presenter and its state.
+        // This means that it won't be recreated on a configuration change, but
+        // it's dangerous as it can produce bugs and leaks. Use it wisely.
+        setRetainInstance(true);
```
Comment (malmstein) [inline todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/taskdetail/TaskDetailFragment.java]: what do you think about using the `Controller` to deal with the lifecycle? Retaining the `Fragment` just to hold a reference to the `Presenter` (which is not View related) seems like a bad idea

File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/tasks/TasksFragment.java [hunk 16]
```diff
@@ -172,7 +177,7 @@ public boolean onOptionsItemSelected(MenuItem item) {
                 mPresenter.loadTasks(true);
                 break;
         }
-        return true;
+        return false;
```
Comment (malmstein) [inline todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/tasks/TasksFragment.java]: why?

File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/tasks/TasksFragment.java [hunk 17]
```diff
@@ -403,12 +415,16 @@ public View getView(int i, View view, ViewGroup viewGroup) {
...
+            if (selectedItem != null && selectedItem.equals(task.getId())) {
+                //noinspection deprecation
+                rowView.setBackgroundDrawable(ContextCompat.getDrawable(
+                        viewGroup.getContext(), R.drawable.selectedTask));
+
             } else {
-                rowView.setBackgroundDrawable(viewGroup.getContext()
-                        .getResources().getDrawable(R.drawable.touch_feedback));
+                //noinspection deprecation
+                rowView.setBackgroundDrawable(ContextCompat.getDrawable(
```
Comment (malmstein) [inline todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/tasks/TasksFragment.java]: it's not part of this PR but I feel like we should be doing a better job at view states than this

File: todoapp/app/src/main/res/values/colors.xml [hunk 18]
```diff
@@ -23,4 +23,6 @@
 
     <drawable name="touchFeedback">#CFD8DC</drawable>
 
+    <drawable name="selectedTask">#CFD8DC</drawable>
```
Comment (malmstein) [inline todoapp/app/src/main/res/values/colors.xml]: no need to duplicate this, how about calling a color `touchStateBackground` or similar and merge `touchFeedback` and `selectedTask`?

File: todoapp/app/src/test/java/com/example/android/architecture/blueprints/todoapp/TasksTabletPresenterTest.java [hunk 19]
```diff
@@ -0,0 +1,152 @@
...
+import org.junit.Test;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Captor;
+import org.mockito.Mock;
+import org.mockito.MockitoAnnotations;
+
+/**
+ * Unit tests for the implementation of {@link TasksTabletPresenter}
+ */
+public class TasksTabletPresenterTest {
```
Comment (malmstein) [inline todoapp/app/src/test/java/com/example/android/architecture/blueprints/todoapp/TasksTabletPresenterTest.java]: the tests turn out very nice!

**Comment by JoseAlcerreca** [comment]
Thanks for your comments @malmstein. Before addressing them one by one, I think the main concern is that I'm storing the presenter in the fragment. This is on purpose since I'm using `setRetainInstance()` to persist  the presenter and its state. Were you aware of that?

File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/TasksMvpController.java [hunk 20]
```diff
@@ -0,0 +1,214 @@
...
+    private TasksPresenter createListPresenter(TasksFragment tasksFragment) {
+        mTasksPresenter = new TasksPresenter(
+                Injection.provideTasksRepository(mFragmentActivity.getApplicationContext()),
+                tasksFragment);
+
+        return mTasksPresenter;
+    }
+
+    @NonNull
+    private TasksFragment findOrCreateTasksFragment() {
```
Comment (saulmm) [inline todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/TasksMvpController.java]: The name of this method is `findOrCreateTaskFragment`, understanding that will implement those operations in the method body (find or create a fragment). 

Actually, is doing one more action not specified in the method signature, which is 'add the fragment to an activity', so I would suggest changing the method's name to  **findOrAddTaskFragment**

**Comment by saulmm** [comment]
After creating a task, if you close the app and open it again, the task created before is not persisted:

https://vid.me/YzdL

**Comment by saulmm** [comment]
The phone design doesn't fit on a large screen in my opinion. I would use a DialogFragment implementing `AddEditTaskContract.View` which is less intrusive and would help to keep the  context.

screen shot 2016-09-25 at 20 43 01

**Comment by JoseAlcerreca** [comment]
On persistence, KI: #123 

On Using a DialogFragment... it's a good idea actually. We're keeping the UI simple but that's easy enough.

I'm still waiting for @malmstein's response about using retained fragments before getting on with the PR.

**Comment by JoseAlcerreca** [comment]
I did a big refactoring that ended up removing a lot of stuff and that's always a good sign :D This is handling state manually now. I syncd with todo-mvp as there were new state-related issues that were fixed and tests added. I think the MvpController class is much better now, please have another look.

File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/TasksMvpController.java (hunk 6 above)
Comment (malmstein) [inline todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/TasksMvpController.java]: I agree with @saulmm on this one!

File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/TasksMvpController.java [hunk 21]
```diff
@@ -0,0 +1,170 @@
...
+import com.example.android.architecture.blueprints.todoapp.tasks.TasksFragment;
+import com.example.android.architecture.blueprints.todoapp.tasks.TasksPresenter;
+import com.example.android.architecture.blueprints.todoapp.util.ActivityUtils;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * Class that creates fragments (MVP views) and makes the necessary connections between them.
+ */
+public class TasksMvpController {
```
Comment (malmstein) [inline todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/TasksMvpController.java]: this is great, nice changes @JoseAlcerreca

File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/taskdetail/TaskDetailPresenter.java [hunk 22]
```diff
@@ -138,4 +143,17 @@ private void showTask(Task task) {
...
         mTaskDetailView.showCompletionStatus(task.isCompleted());
     }
+
+    @Nullable
+    public String getTaskId() {
+        return mTaskId;
+    }
+
+    @Override
+    public void setTaskId(@Nullable String taskId) {
```
Comment (malmstein) [inline todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/taskdetail/TaskDetailPresenter.java]: when are we calling this as `null`?

File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/tasks/TasksActivity.java [hunk 23]
```diff
@@ -62,30 +63,28 @@ protected void onCreate(Bundle savedInstanceState) {
...
+        tasksMvpTabletController = TasksMvpController.createTasksView(this, taskId);
+        if (currentFiltering != null) {
+            tasksMvpTabletController.setFiltering(currentFiltering);
         }
     }
 
     @Override
     public void onSaveInstanceState(Bundle outState) {
-        outState.putSerializable(CURRENT_FILTERING_KEY, mTasksPresenter.getFiltering());
+        outState.putSerializable(CURRENT_FILTERING_KEY,
```
Comment (malmstein) [inline todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/tasks/TasksActivity.java]: 👍

File: todoapp/app/src/main/res/values-sw600dp/values.xml [hunk 24]
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
+    <bool name="isTablet">true</bool>
```
Comment (malmstein) [inline todoapp/app/src/main/res/values-sw600dp/values.xml]: as stated above, can we get this value from the `MVPController`?

File: todoapp/app/src/main/res/values/colors.xml (hunk 18 above)
Comment (malmstein) [inline todoapp/app/src/main/res/values/colors.xml]: ping

File: todoapp/build.gradle [hunk 25]
```diff
@@ -29,13 +29,13 @@ ext {
     buildToolsVersion = '23.0.2'
 
     // App dependencies
-    supportLibraryVersion = '23.4.0'
+    supportLibraryVersion = '24.1.1'
```
Comment (malmstein) [inline todoapp/build.gradle]: if we are doing this here, let's make sure a new PR is open once this is merged when we update all the libraries

File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/TasksMvpController.java (hunk 21 above)
Comment (JoseAlcerreca) [inline todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/TasksMvpController.java]: 🕶

File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/TasksMvpController.java (hunk 6 above)
Comment (JoseAlcerreca) [inline todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/TasksMvpController.java]: Done

File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/tasks/TasksFragment.java (hunk 17 above)
Comment (JoseAlcerreca) [inline todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/tasks/TasksFragment.java]: agreed

File: todoapp/app/src/main/res/values-sw600dp/values.xml (hunk 24 above)
Comment (JoseAlcerreca) [inline todoapp/app/src/main/res/values-sw600dp/values.xml]: I think you mean from `ActivityUtils`.

**Comment by JoseAlcerreca** [comment]
PTAL and let's merge. 

TODO:
 - DialogFragment
 - README

(we're still in dev-)
